packages/editor/src/components/rich-text/format-toolbar/link-container.js

```diff
@@ -19,95 +23,200 @@ import { filterURLForDisplay } from '../../../utils/url';
 
 const stopKeyPropagation = ( event ) => event.stopPropagation();
 
-const LinkContainer = ( props ) => {
-	const {
-		editLink,
-		formats,
-		isEditing,
-		isPreviewing,
-		linkValue,
-		onChangeLinkValue,
-		onKeyDown,
-		opensInNewWindow,
-		selectedNodeId,
-		setLinkTarget,
-		settingsVisible,
-		submitLink,
-		toggleLinkSettingsVisibility,
-	} = props;
-
-	const linkSettings = settingsVisible && (
-		<div className="editor-format-toolbar__link-modal-line editor-format-toolbar__link-settings">
-			<ToggleControl
-				label={ __( 'Open in New Window' ) }
-				checked={ opensInNewWindow }
-				onChange={ setLinkTarget } />
-		</div>
-	);
-
-	return (
-		<Fill name="RichText.Siblings">
-			<PositionedAtSelection className="editor-format-toolbar__link-container">
-				<Popover
-					position="bottom center"
-					focusOnMount={ isEditing ? 'firstElement' : false }
-					key={ selectedNodeId /* Used to force rerender on change */ }
+class LinkContainer extends Component {
+	constructor( { link, selection } ) {
+		super( ...arguments );
+
+		this.editLink = this.editLink.bind( this );
+		this.submitLink = this.submitLink.bind( this );
+		this.onKeyDown = this.onKeyDown.bind( this );
+		this.onChangeInputValue = this.onChangeInputValue.bind( this );
+		this.toggleLinkSettingsVisibility = this.toggleLinkSettingsVisibility.bind( this );
+		this.setLinkTarget = this.setLinkTarget.bind( this );
+
+		this.state = {
+			inputValue: link && link.attributes && link.attributes.href ? null : '',
+			opensInNewWindow: null,
+			lastSelection: selection,
+			selectionKey: 0,
+		};
+	}
+
+	static getDerivedStateFromProps( { link, selection }, state ) {
+		if ( selection === state.lastSelection ) {
+			return null;
+		}
+
+		return {
+			inputValue: link.attributes && link.attributes.href ? null : '',
+			opensInNewWindow: null,
+			lastSelection: selection,
+			selectionKey: state.selectionKey + 1,
+		};
+	}
+
+	onKeyDown( event ) {
+		if ( event.keyCode === ESCAPE ) {
+			event.stopPropagation();
+			this.props.removeFormat( 'a' );
+			this.setState( {
+				inputValue: null,
+				opensInNewWindow: null,
+			} );
+		}
+
+		if ( [ LEFT, DOWN, RIGHT, UP, BACKSPACE, ENTER ].indexOf( event.keyCode ) > -1 ) {
+			// Stop the key event from propagating up to maybeStartTyping in BlockListBlock.
+			event.stopPropagation();
+		}
+	}
+
+	onChangeInputValue( inputValue ) {
+		this.setState( { inputValue } );
+	}
+
+	toggleLinkSettingsVisibility() {
+		this.setState( ( state ) => ( { settingsVisible: ! state.settingsVisible } ) );
```

Comment:
You want to have it predictable because `setState` might be executed in a batch. That's why they recommend using a function to ensure you always use the explicitly passed previous value.

Reply:
Ah okay, cool. A comment here mentioning that would be handy, but that works for me.